src/main/java/org/zalando/nakadi/validation/SchemaCompatibilityChecker.java

```diff
@@ -0,0 +1,116 @@
+package org.zalando.nakadi.validation;
+
+import com.google.common.collect.Lists;
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.validation.schema.NotSchemaConstraint;
+import org.zalando.nakadi.validation.schema.PatternPropertiesConstraint;
+import org.zalando.nakadi.validation.schema.SchemaConstraint;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Stack;
+
+@Component
+public class SchemaCompatibilityChecker {
+
+    final private List<SchemaConstraint> CONSTRAINTS = Lists.newArrayList();
+
+    public SchemaCompatibilityChecker() {
+        CONSTRAINTS.add(new NotSchemaConstraint());
+        CONSTRAINTS.add(new PatternPropertiesConstraint("patternProperties"));
+    }
+
+    public List<SchemaIncompatibility> checkConstraints(final Schema schema) {
+        final List<SchemaIncompatibility> incompatibilities = new ArrayList<SchemaIncompatibility>();
+
+        recursiveCheckConstraints(schema, new Stack<String>(), incompatibilities);
+
+        return incompatibilities;
+    }
+
+    private void recursiveCheckConstraints(
+            final Schema schema,
+            final Stack<String> jsonPath,
+            final List<SchemaIncompatibility> schemaIncompatibilities) {
+
+        for (final SchemaConstraint constraint : CONSTRAINTS) {
```

Comment:
I lost the interest because of the lack of functional data structures. For example, list manipulation in functional languages is a breeze, but in Java it's ackward. I even decided to use a `Stack` to keep recursion state =/